sopel/modules/currency.py

```diff
@@ -1,102 +1,189 @@
+        return rate
+
+
+def update_rates(bot):
+    global rates_fiat_json
+    global rates_btc_json
+
+    # If we have data that are less than 24h old, return
+    if 'date' in rates_fiat_json:
+        if time.time() - rates_fiat_json['date'] < 24 * 60 * 60:
+            return
+
+    # Update crypto rates
+    request = requests.get(crypto_url)
+    request.raise_for_status()
+    rates_btc_json = request.json()
+
+    # Update fiat rates
+    if bot.config.currency.fixer_io_key is not None:
+        request = requests.get(fixer_url.format(bot.config.currency.fixer_io_key))
+        if not request.json()['success']:
+            stderr(str(request.json()['error']))
+            return bot.reply('Sorry, something went wrong')
+    else:
+        request = requests.get(fiat_url)
+
+    request.raise_for_status()
+    rates_fiat_json = request.json()
+    rates_fiat_json['date'] = time.time()
+    rates_fiat_json['rates']['EUR'] = 1.0  # Put this here to make logic easier
+
+
+def get_rate(of, to):
+    global rates_fiat_json
+    of = of.upper()
+    to = to.upper()
+
+    if of == 'BTC' or to == 'BTC':
+        if of == 'BTC':
+            code = to
+            reverse = False
+        else:
+            code = of
+            reverse = True
+
+        return btc_rate(code, reverse)
+
+    if of not in rates_fiat_json['rates']:
+        return "Sorry {} isn't currently supported".format(of)
+
+    if to not in rates_fiat_json['rates']:
+        return "Sorry {} isn't currently supported".format(to)
+
+    return (1 / rates_fiat_json['rates'][of]) * rates_fiat_json['rates'][to]
+
+
+def exchange(bot, match):
     """Show the exchange rate between two currencies"""
-    if not trigger.group(2):
-        return bot.reply("No search term. An example: .cur 20 EUR in USD")
-    match = regex.match(trigger.group(2))
+
+    # match = regex.match(trigger)                                                      # that
     if not match:
-        # It's apologetic, because it's using Canadian data.
         bot.reply("Sorry, I didn't understand the input.")
         return NOLIMIT
 
-    amount, of, to = match.groups()
+    update_rates(bot)  # Try and update rates. Rate-limiting is done in update_rates()
+
+    query = match.string
+
+    others = query.split()
+    amount = others.pop(0)
+    of = others.pop(0)
+    others.pop(0)
+
+#    amount, of, _, *others = query.split() # I'd much rather use this, but it's not python 2.7 compatible
+
     try:
         amount = float(amount)
     except ValueError:
         bot.reply("Sorry, I didn't understand the input.")
     except OverflowError:
         bot.reply("Sorry, input amount was out of range.")
-    display(bot, amount, of, to)
+
+    out_string = '{} {} is'.format(amount, of.upper())
+    for to in others:
+        try:
+            out_string = build_reply(bot, amount, of.upper(), to.upper(), out_string)
+        except (requests.exceptions.HTTPError, requests.exceptions.ConnectionError) as e:
```

Comment:
`build_reply` doesn't seem to throw any of these exceptions as it doesn't actually make any requests in the call stack.

This block would better serve in `exchange` where I left another comment about error handling.

Reply:
There are also more potential exceptions that could happen than just these two types. See [this `help` code](https://github.com/sopel-irc/sopel/blob/eb23031ea7729836de6eaece61b42434a8b44fd5/sopel/modules/help.py#L40-L48) for one solid way to handle everything at once. (I don't want to have `sopel.web` go back to wrapping `requests`, but… they could really use it. The error-catching boilerplate can be tiresome.)

Reply:
I looked at the request module and some discussions about it. it seems `requests.exceptions.RequestException` wraps up all of the mentioned exceptions, so I just catch this now.

Reply:
Also, I moved this to the exchange function.

Reply:
You "moved" this? Then why is it still here? :laughing:

Reply:
That is a great question. must've forgotten to save the file before committing or something :laughing: